roles/matrix-bridge-appservice-discord/tasks/setup_install.yml

```diff
@@ -1,5 +1,28 @@
 ---
 
+- block:
+  - set_fact:
+      matrix_appservice_discord_sqlite_db_path: "{{ matrix_appservice_discord_data_path }}/{{ matrix_appservice_discord_database_filename_name }}"
+
+  - name: Check if an SQLite database already exists
+    stat:
+      path: "{{ matrix_appservice_discord_sqlite_db_path }}"
+    register: matrix_appservice_discord_sqlite_db_path_stat_result
+
```

Comment:
Maybe we could do sth here with `import_tasks` for an automigration

Reply:
Auto-migration would be ideal. At first I started on that, but then I figured it might be somewhat dangerous.

For example, importing the Appservice Discord database **always** gives me at least a few errors ("... cannot have multiple primary keys in `emoji` table", etc). As of yet, I'm still not sure how perfectly that migration went.

So it makes me see it as something a bit dangerous and a somewhat big change, that users should do explicitly (similar to how we don't auto-upgrade between Postgres versions -- although that's even more dangerous).

Doing auto-migration also seemed more difficult to do, as it involves:

- (*already done*) detecting whether we need to migrate
- stopping the bridge service, so that we can work with the database exclusively
- doing the migration via `import_tasks` (should be easy). This likely involves doing the same `file.db` -> `file.db.backup` rename
- then doing all the configuration tasks -- we need to rebuild a bridge's `config.yml` after all, as there's no `file.db` database anymore and it's all in Postgres already. If the bridge were to be started at this moment, it would create a new empty database (likely not too catastrophic though).
- then starting the bridge again if it had been stopped in step 2, above. Is this to be done in the role itself or postponed until the very end of playbook execution (`matrix-common-after`)? If someone does just `--tags=setup-appservice-discord`, then we won't even make it to where we expect to in `matrix-common-after`. Well, that's rare, but ..

It should be possible to have auto-migration though. What we have now is a little annoying to use.

Reply:
An alternative to both auto-migration and erroring-out like we do now is to: detect that there's leftover SQLite data and automatically switch to using SQLite when generating the configuration.

We could leave the `matrix_postgres_additional_databases` definitions in place and have these empty databases get created and left unused.

Whenever we detect that leftover data and auto-switch to SQLite, we could push a warning (notice) message to be displayed at the end of playbook execution. We do such a thing for major Postgres upgrades. We automatically keep people on the old version (instead of trying to auto-upgrade them), and we tell them with a message, letting them run the upgrade whenever they feel comfortable. Code is here: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/6e1dfb62f0d371fdb82fc7a9e635dd32d0179e86/roles/matrix-postgres/tasks/setup_postgres.yml#L21-L31

What do you think? To me, it sounds reasonable and simple to implement.. It's also not dangerous.

Reply:
That adds an extra layer of "let the user choose", and i think it becomes harder and harder to keep the change small. ideally, we would automigrate everything and have the roles themselves with sqlite as default, supporting delaying, external postgres, etc etc blows up the scope, and users would except that a merged feature was tested with every combination.
TL;DR: could do, but might bite us later, and we should encourage users to migrate to postgres by making the path to migrate easy (encourage users to actually do it and not perpetually delay it)